### PR TITLE
Make the dx12 backend ocean ready 🌊 

### PR DESCRIPTION
**Pull request by msiglreith**
* Remove the `WRITE_COMBINE` memory property and adjust it towards the vulkan model
* Implement uniform and storage buffers for dx12 where we dynamically create view objects
* Fix exposed memory heaps and types (query available memory of the system)

### CODE REVIEW

File: examples/hal/quad/main.rs [hunk 1]
```diff
@@ -314,7 +314,7 @@ fn main() {
         memory_types.iter().find(|mem_type| {
             buffer_req.type_mask & (1 << mem_type.id) != 0 &&
             mem_type.properties.contains(m::CPU_VISIBLE) &&
-            mem_type.properties.contains(m::WRITE_COMBINED)
+            mem_type.properties.contains(m::COHERENT)
```
Comment (kvark) [inline examples/hal/quad/main.rs]: AFAIK, technically COHERENT doesn't imply either upload/download types. It's orthogonal, and thus we shouldn't request it here.

File: src/backend/dx12/src/device.rs [hunk 2]
```diff
@@ -339,12 +340,44 @@ impl Device {
 
     fn view_image_as_storage(
         &mut self,
-        _resource: *mut winapi::ID3D12Resource,
-        _kind: image::Kind,
-        _format: winapi::DXGI_FORMAT,
-        _range: &image::SubresourceRange,
+        resource: *mut winapi::ID3D12Resource,
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: nice!

File: src/backend/dx12/src/device.rs [hunk 3]
```diff
@@ -359,9 +392,17 @@ impl d::Device<B> for Device {
...
+            CPUPageProperty: self.heap_properties[mem_type.id].page_property,
+            MemoryPoolPreference: self.heap_properties[mem_type.id].memory_pool,
+            CreationNodeMask: 0,
+            VisibleNodeMask: 0,
+        };
+
         let desc = winapi::D3D12_HEAP_DESC {
             SizeInBytes: size,
-            Properties: conv::map_heap_properties(mem_type.properties),
+            Properties: properties,
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: oh, interesting

File: src/backend/dx12/src/device.rs [hunk 4]
```diff
@@ -1318,12 +1354,92 @@ impl d::Device<B> for Device {
...
+                pso::DescriptorWrite::UniformBuffer(ref views) |
+                pso::DescriptorWrite::StorageBuffer(ref views) => {
+                    num_views += views.len();
+                },
+                _ => (),
+            }
+        }
+
+        let mut buffer_heap = n::DescriptorCpuPool {
+            heap: Self::create_descriptor_heap_impl(
```
Comment (kvark) [inline src/backend/dx12/src/device.rs]: wait, we create a temporary heap upon updating descriptor sets? that sound heavy. Could you raise the topic on an issue/gitter please?

File: src/backend/dx12/src/lib.rs [hunk 5]
```diff
@@ -113,26 +175,88 @@ impl core::Adapter<Backend> for Adapter {
...
-                properties: memory::CPU_VISIBLE | memory::CPU_CACHED,
-                heap_index: 0,
-            },
-            core::MemoryType {
-                id: 2,
-                properties: memory::CPU_VISIBLE | memory::WRITE_COMBINED,
-                heap_index: 0,
-            },
-        ];
+        let base_memory_types = match device.private_caps.memory_architecture {
```
Comment (kvark) [inline src/backend/dx12/src/lib.rs]: very nice!

File: src/backend/dx12/src/lib.rs [hunk 6]
```diff
@@ -153,13 +277,34 @@ impl core::Adapter<Backend> for Adapter {
...
+                    segment,
+                    &mut mem_info,
+                ));
+                mem_info.Budget
+            };
+
+            let local = query_memory(winapi::DXGI_MEMORY_SEGMENT_GROUP_LOCAL);
+            match device.private_caps.memory_architecture {
+                MemoryArchitecture::NUMA => {
+                    let non_local = query_memory(winapi::DXGI_MEMORY_SEGMENT_GROUP_NON_LOCAL);
```
Comment (kvark) [inline src/backend/dx12/src/lib.rs]: excellent 👍

File: examples/hal/quad/main.rs (hunk 1 above)
Comment (msiglreith) [reply examples/hal/quad/main.rs]: Oh yeah, well, we don't expose invalidation and flushing, so COHERENT is a bit useless in our API imo /:

**Comment by msiglreith** [comment]
Rebased upon master and removed the `COHERENT` flag for now.

**Pull request merged**